io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java

```diff
@@ -0,0 +1,307 @@
+                    SSOAgentConstants.ErrorMessages.ID_TOKEN_PARSE.getCode(), e);
+        }
+    }
+
+    private void handleErrorTokenResponse(TokenRequest tokenRequest, TokenResponse tokenResponse) {
+
+        TokenErrorResponse errorResponse = tokenResponse.toErrorResponse();
+        JSONObject requestObject = requestToJson(tokenRequest);
+        JSONObject responseObject = errorResponse.toJSONObject();
+        logger.log(Level.INFO, "Request object for the error response: ", requestObject);
+        logger.log(Level.INFO, "Error response object: ", responseObject);
+    }
+
+    private void handleErrorAuthorizationResponse(AuthorizationResponse authzResponse) {
+
+        AuthorizationErrorResponse errorResponse = authzResponse.toErrorResponse();
+        JSONObject responseObject = errorResponse.getErrorObject().toJSONObject();
+        logger.log(Level.INFO, "Error response object: ", responseObject);
+    }
+
+    private TokenResponse getTokenResponse(TokenRequest tokenRequest) {
+
+        TokenResponse tokenResponse = null;
+        try {
+            tokenResponse = TokenResponse.parse(tokenRequest.toHTTPRequest().send());
+        } catch (com.nimbusds.oauth2.sdk.ParseException | IOException e) {
+            logger.log(Level.ERROR, "Error while parsing token response.", e);
+        }
+        return tokenResponse;
+    }
+
+    private TokenRequest getTokenRequest(AuthorizationCode authorizationCode) {
+
+        URI callbackURI = oidcAgentConfig.getCallbackUrl();
+        AuthorizationGrant authorizationGrant = new AuthorizationCodeGrant(authorizationCode, callbackURI);
+        ClientID clientID = oidcAgentConfig.getConsumerKey();
+        Secret clientSecret = oidcAgentConfig.getConsumerSecret();
+        ClientAuthentication clientAuthentication = new ClientSecretBasic(clientID, clientSecret);
+        URI tokenEndpoint = oidcAgentConfig.getTokenEndpoint();
+
+        return new TokenRequest(tokenEndpoint, clientAuthentication, authorizationGrant);
+    }
+
+    private JSONObject requestToJson(AbstractRequest request) {
+
+        JSONObject obj = new JSONObject();
+        obj.appendField("tokenEndpoint", request.toHTTPRequest().getURI().toString());
+        obj.appendField("request body", request.toHTTPRequest().getQueryParameters());
+        return obj;
+    }
+
+    private Map<String, Object> getUserAttributes(String idToken) throws SSOAgentServerException {
+
+        Map<String, Object> userClaimValueMap = new HashMap<>();
+        try {
+            JWTClaimsSet claimsSet = SignedJWT.parse(idToken).getJWTClaimsSet();
+            Map<String, Object> customClaimValueMap = claimsSet.getClaims();
+
+            for (String claim : customClaimValueMap.keySet()) {
+                if (!SSOAgentConstants.OIDC_METADATA_CLAIMS.contains(claim)) {
+                    userClaimValueMap.put(claim, customClaimValueMap.get(claim));
```

Comment:
If `customClaimValueMap.get(claim)` always returns a String can't we specify the data types of the map as `Map<String, String>`?

Reply:
Only the keys will be String format. Values can have different types (JSONArray, Arraylist, Date, etc.)